### PR TITLE
Feature/cdap 947 expose format in mr

### PR DESCRIPTION
**Pull request by albertshau**
Exposing format and schema in the api, and adding ways in mapreduce to use them for reading from streams. Mostly moving classes and import changes. Actual code changes are in StreamInputFormat, StreamBatchReadable, MapreduceRuntimeService, and FormatStreamEventDecoder.

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/data/schema/package-info.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Defines schema for use in programs and data exploration.
+ */
+package co.cask.cdap.api.data.schema;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/data/schema/package-info.java]: Are you planning to move this https://github.com/caskdata/cdap/blob/develop/cdap-api/src/main/java/co/cask/cdap/internal/io/package-info.java into here as well?

File: cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java [hunk 2]
```diff
@@ -243,4 +295,27 @@ public URI toURI() {
...
+  private static FormatSpecification decodeFormatSpec(String encodedFormatSpecification) {
+    if (encodedFormatSpecification == null) {
+      return null;
+    }
+    try {
+      String decodedFormatSpecification = URLDecoder.decode(encodedFormatSpecification, "UTF-8");
+      return GSON.fromJson(decodedFormatSpecification, FormatSpecification.class);
+    } catch (UnsupportedEncodingException e) {
+      // shouldn't ever happen
+      return null;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java]: Although it won't happen, it'd be good to have unify way to handle exception (you propagate above but return null in here).

File: cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java [hunk 3]
```diff
@@ -226,6 +271,13 @@ public String getDecoderType() {
   }
 
   /**
+   * Returns the {@link FormatSpecification} for reading the body of the stream.
+   */
+  public FormatSpecification getBodyFormatSpec() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java]: Better name the method as `getFormatSpecification`.

File: cdap-api/src/main/java/co/cask/cdap/api/data/schema/package-info.java (hunk 1 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/data/schema/package-info.java]: ah yes should have copied it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 4]
```diff
@@ -504,13 +506,20 @@ private void configureStreamInput(Job job, StreamBatchReadable stream) throws IO
     StreamInputFormat.setStreamPath(job, streamPath.toURI());
     StreamInputFormat.setTimeRange(job, stream.getStartTime(), stream.getEndTime());
 
-    String decoderType = stream.getDecoderType();
-    if (decoderType == null) {
-      // If the user don't specify the decoder, detect the type from Mapper/Reducer
-      setStreamEventDecoder(job);
+    FormatSpecification formatSpecification = stream.getBodyFormatSpec();
+    if (formatSpecification != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Can you open a JIRA for the case when the format spec is available but the mapper/reducer is using a different type, we should be able to do type projection into the target type.

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT472

File: cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java (hunk 2 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java]: agreed, will propagate it.

File: cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java (hunk 3 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java]: done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 4 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: created https://issues.cask.co/browse/CDAP-1149, will put it in the comment as well

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/decoder/FormatStreamEventDecoder.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * Create a decoder for stream events that decodes the body of the stream using the given initialized format.
+   *
+   * @param bodyFormat Initialized format.
+   */
+  public FormatStreamEventDecoder(RecordFormat<ByteBuffer, StructuredRecord> bodyFormat) {
+    this.bodyFormat = bodyFormat;
+    this.bodySchema = bodyFormat.getSchema();
+    List<Schema.Field> fields = Lists.newArrayList(
+      Schema.Field.of(TIMESTAMP_FIELD, Schema.of(Schema.Type.LONG)),
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/decoder/FormatStreamEventDecoder.java]: Why need to repeat the timestamp in the record if it is already in the mapper key?

**Comment by chtyim** [comment]
Just one comment, rest LGTM.

**Pull request merged**